compiler/rustc_metadata/src/rmeta/encoder.rs

```diff
@@ -1452,8 +1441,9 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
                     self.encode_info_for_assoc_item(def_id);
                 }
             }
-            if let DefKind::Coroutine = def_kind {
-                let data = self.tcx.coroutine_kind(def_id).unwrap();
+            if def_kind == DefKind::Closure
+                && let Some(data) = self.tcx.coroutine_kind(def_id)
```

Comment:
Does `coroutine_kind`/`is_coroutine` panic on non-closure def ids?
If it just returns `None`/`false` in those cases then the `def_kind == DefKind::Closure` conditions are redundant in a few places in this file.

Reply:
Correction: it's redundant only if `def_kind` isn't already calculated, otherwise it may be important for incremental performance to avoid additional query dependencies.

Reply:
> Does coroutine_kind/is_coroutine panic on non-closure def ids?

Indeed, if you remove the precondition ﻿`def_kind == DefKind::Closure` from this location, it will cause a panic. The panic is triggered by the `﻿tcx.hir().get_by_def_id(def_id)`.

However, it doesn't mean that all `﻿is_coroutine`/`﻿coroutine_kind` require this precondition. I will attempt to reduce this reliance in the upcoming patch.